src/audio/kpb.c

```diff
@@ -150,14 +151,14 @@ static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
 	}
 
 	/* Initialize draining task */
-	schedule_task_init(&kpb->draining_task, /* task structure */
-			   SOF_SCHEDULE_EDF, /* utilize EDF scheduler */
-			   SOF_TASK_PRI_ALMOST_IDLE, /* almost idle priority */
-			   kpb_draining_task, /* task function */
-			   NULL, /* no complete function */
-			   &kpb->draining_task_data, /* task private data */
-			   0, /* core on which we should run */
-			   SOF_SCHEDULE_FLAG_IDLE);
+	schedule_edf_task_init(&kpb->draining_task, /* task structure */
+			       /* almost idle priority */
+			       SOF_TASK_PRI_ALMOST_IDLE,
+			       kpb_draining_task, /* task function */
+			       NULL, /* no complete function */
+			       &kpb->draining_task_data, /* task private data */
+			       0, /* core on which we should run */
+			       SOF_SCHEDULE_FLAG_IDLE);
```

Comment:
Almost all of the rest of the generic scheduling operations are still valid. I would say initializing specific scheduling tasks with never used parameters are pointless and leads only to confusion. You need to specify task type anyway, so using dedicated task init function is a small price for improved clarity.

Reply:
@tlauda sorry, I still disagree. I think this is rather common, that some implementations of a generic function don't use all its parameters, nothing terribly confusing there. Moreover, in such implementations you can mark those unused arguments as __attribute__((unused)) to avoid compiler warnings and simultaneously to document, that that is on purpose.

Reply:
It's just not practical in this case and doesn't benefit us at all. You still need to decide what type of task you want to create. It's the same mechanism as in the object oriented programming languages, where you create specific object and use it based on its generic type. Giving that EDF scheduler is still growing and will require e.g. more operations then simple run method, it is highly unpractical to enforce other schedulers to take task_ops struct.

Reply:
@tlauda I think this is similar to scheduling policies in Linux / POSIX. You have a generic scheduling API and you select which scheduling policy you want to govern your task, which will select the respective algorithm. I understand that in some cases generic APIs are a hindrance, they make you drag all the numerous users of that API with you for each change. E.g. that was the case when I was reworking the interrupt handling in SOF. So, this my comment obviously isn't a blocker, if everyone else is fine with this. Particularly this might make sense if you have plans to further work on the EDF scheduler and on the API. Maybe once it's stabilised we can re-unify them.

Reply:
@tlauda @lyakh I'd lean towards having the separate create functions. As long as the rest of the API is aligned (across schedulers), having a different create function seems ok. Based on @tlauda your earlier comment, this is the intent.
 
OTOH, in Linux kernel, the EDF scheduler is exposed via the standard API (sched_setscheduler takes a sched_param struct that has additional fields for EDF) so it is posssible to use the same interface as well.

Reply:
Overall I would like to see consistent APIs for scheduling and all core RTOS services across SOF. Yes, I do see the point to reduce the number of arguments (that are optional for some users while mandatory for others), but keeping APIs consistent guarantees simpler service/code reuse for future use case support (that have some differentiation to current services).
I have no problem here if schedule_task_init_edf() and schedule_task_init_ll() (note the naming change) are helper APIs that simply populate struct task before calling generic schedule_task_init().